src/components/utils/src/threads/thread_win.cc

```diff
@@ -61,46 +61,62 @@ size_t Thread::kMinStackSize = 0;
 void Thread::cleanup(void* arg) {
   SDL_AUTO_TRACE();
   Thread* thread = static_cast<Thread*>(arg);
-  thread->isThreadRunning_ = false;
+  thread->thread_state_ = kThreadStateNone;
+  thread->thread_command_ = kThreadCommandNone;
   thread->state_cond_.Broadcast();
 }
 
 void* Thread::threadFunc(void* arg) {
-  // 0 - state_lock unlocked
-  //     stopped   = 0
-  //     running   = 0
-  //     finalized = 0
-  // 4 - state_lock unlocked
-  //     stopped = 1
-  //     running = 1
-  //     finalized = 0
-  // 5 - state_lock unlocked
-  //     stopped = 1
-  //     running = 1
-  //     finalized = 1
   SDL_DEBUG("Thread #" << GetCurrentThreadId() << " started successfully");
 
   threads::Thread* thread = static_cast<Thread*>(arg);
   DCHECK(thread);
 
   thread->state_lock_.Acquire();
+  thread->thread_state_ = kThreadStateIdle;
   thread->state_cond_.Broadcast();
 
-  while (!thread->finalized_) {
+  // We have special variable for controlling iterations/exiting thread
+  // in order to separate decision logic (continue iterations or exit?)
+  // from controlling while cycle
+  bool continueIterations = true;
+
+  while (continueIterations) {
     SDL_DEBUG("Thread #" << GetCurrentThreadId() << " iteration");
-    thread->run_cond_.Wait(thread->state_lock_);
+    while (kThreadCommandNone == thread->thread_command_) {
+      thread->run_cond_.Wait(thread->state_lock_);
+    }
     SDL_DEBUG("Thread #" << GetCurrentThreadId() << " execute. "
                          << "stopped_ = " << thread->stopped_
                          << "; finalized_ = " << thread->finalized_);
-    if (!thread->stopped_ && !thread->finalized_) {
-      thread->isThreadRunning_ = true;
 
-      thread->state_lock_.Release();
-      thread->delegate_->threadMain();
-      thread->state_lock_.Acquire();
-
-      thread->isThreadRunning_ = false;
+    switch (thread->thread_command_) {
+      case kThreadCommandRun:
+        thread->thread_state_ = kThreadStateRunning;
+        pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL);
```

Comment:
@pestOO, really so far I do/test only linux/posix flavour on my machine but I contacted @LuxoftAKutsan for how-to info/set-up.